### DISCUSSION TITLE
RFC: Customisation features

### DISCUSSION THREAD

**Discussion by joulev** [joulev/ezkomment#145] [General]
Today I played with the customisation a bit and eventually came up with the [`test-customisation-feature`](https://github.com/joulev/ezkomment/tree/test-customisation-feature) branch. You can test it by checking out the branch, starting the Next.js server and going to http://localhost:3000/test/index.html.

`/api/temp` is an endpoint that accepts `GET` and `POST` for comments. Right now it's a dummy endpoint for simplicity's sake, in the actual app we will use an actual endpoint.

So, about the [index.html](https://github.com/joulev/ezkomment/blob/test-customisation-feature/public/test/index.html) file, all of that can be customised by the user, except the ending `<script>` tag which must be there for the whole thing to work. (See the file for more details.)

Now this is pretty messy and will require cleaning up/refactoring a bit, also there might be some bugs here and there that need attention, but this is how it basically looks like. The user will simply add a HTML file, from `<!DOCTYPE html>` onwards, and we post-process that HTML string by adding the script tag at the end. I think that will work for the customisation features.

However, we need to think a bit more if we do plan to support editing. Probably something like

```html
<!-- hide the <span> accordingly -->
<button type="submit">
  <span data-ezk="post-mode">Post</span>
  <span data-ezk="edit-mode">Edit</span>
</button>
```

Or we should expose some variables, say, `isEditMode`, so the user can manipulate the DOM with his own code.

Et cetera.

This is a request for comment: what do you think?

---

**Edit 25 June 17:19 SGT:** About Markdown and XSS: now they are resolved too. Now we have two ways to render Markdown:

* Render on the client: Using `markdown-it` (0047e4bed0489075548d34ee4f885da0ef0b6131)
  * Pros: Can add comment preview if we want
  * Cons: Larger bundle size
* Render on the server: Using the `remark`-`rehype` packages (ec21301cdf731a0b7962d3da3587aa146c533385). Note: this is how we are rendering the posts and documentation pages.
  * Pros: Zero overhead on the client
  * Cons: You can't have comment preview

Both are safe against XSS.

<img width="666" alt="image" src="https://user-images.githubusercontent.com/44609036/175766797-f76b96c0-eeb2-4c49-af4b-963cccafc8c2.png">

<img width="436" alt="image" src="https://user-images.githubusercontent.com/44609036/175766952-58777678-244b-4a03-b07a-4bc8a290a377.png">

I think both are fine... comment previews are not that necessary (who knows Markdown doesn't need it and who doesn't know Markdown doesn't need it either), but `markdown-it` is also not that big. If any, I slightly prefer the `remark`-`rehype` ecosystem since it more correlates with how we are actually rendering Markdown in the whole app.

**Comment by VietAnh1010** [comment]
> hide the <span> accordingly

I don't quite get the purpose of this. Probably because we don't want the user to edit the span?

> Or we should expose some variables, say, isEditMode, so the user can manipulate the DOM with his own code.

If `isEditMode` is `true`, the user can grab the DOM and manipulate it with his code? I think this is necessary.

For the user to test his configuration, maybe we can let the code run in a sandbox environment, similar to [codesandbox](https://codesandbox.io). If there is an error, we will reject the user's code.

**Reply by joulev** [reply]
> I don't quite get the purpose of this

The button should change the label according to whether the commenter is in edit mode or not. So in edit mode we hide all `[data-ezk="post-mode"]` elements, while in post mode we hide all `[data-ezk="edit-mode"]` elements with JavaScript.

> the user can grab the DOM and manipulate it with his code?

From *the very start* the user can manipulate the DOM any way he likes. The question is that we expose some sort of state variables so the user knows how to update the DOM for each state.

> For the user to test his configuration, maybe we can let the code run in a sandbox environment, similar to [codesandbox](https://codesandbox.io/). If there is an error, we will reject the user's code.

We already have a customise page for each site where the code is rendered live &ndash; the user can use that. More complex solutions are not necessary; it's his code for his site, if his code is bad only he suffers.

**Reply by VietAnh1010** [reply]
> However, we need to think a bit more if we do plan to support editing

I interpret `editing` wrongly lmao.
So basically the problem now is that we don't have a good design (yet) for editing comments.

> We already have a customized page for each site where the code is rendered live – the user can use that. More complex solutions are not necessary; it's his code for his site, if his code is bad only he suffers.

Just check the app, indeed the change is rendered live. 
I changed the CSS file and nothing happen. Should have edited the HTML file instead to see the immediate re-rendering.

> Render on the client: Using markdown-it

Not sure about this at the moment, but I will prefer to be able to preview my comment.

> From the very start the user can manipulate the DOM any way he likes. The question is that we expose some sort of state variables so the user knows how to update the DOM for each state.

So, the variables are hidden, and if `isEditMode` then we expose some variables &ndash; the user can access these variables and changes them accordingly to see the effects, right? This is a good feature.

**Reply by joulev** [reply]
> So, the variables are hidden

Technically this is not correct, since the app is open source and no matter how much we hide it it's still there and accessible in client browser. However yes, we only make documentation and ensure backward-compatibility for these exposed states. Maybe we will prefix all other variables with `__` or something to indicate "privateness".

> the user can access these variables and changes them accordingly

Hmm not exactly what I meant. The user can read the variables and update the DOM according to the value. But to let the user update the variables is not a good idea IMO, lots of unpredictable things can happen.

**Reply by joulev** [reply]
Hmm, I think for now, when editing is not planned, we don't need to think over all this logic yet. Without editing it becomes pretty simple and I think the spec can already be finalised, I'm writing it.

Even with editing I think exposing some variables is enough, the core idea of the feature is to give users full customisation power anyway. No need of hiding `<span>` conditionally.

**Reply by VietAnh1010** [reply]
Agree, if possible we will give the user all the power they want.

Hopefully they won't shoot themselves in the foot (like programming in C).

**Reply by joulev** [reply]
> (like programming in C).

Or CSS. F*ck CSS. Life is too short to be a CSS dev.

**Reply by VietAnh1010** [reply]
> Life is too short to be a CSS dev.

Or, according to Fireship, 2022: "Life is too short to be coding in Java man". (Somehow Java is still bearable to me, and hope that Java can implement ADTs so I can code in Java in a similar pattern as Haskell).

Joking aside, I think we should prefix the variables with `_`. Easy enough to remember (I don't really like double underscores in JS/TS).

**Reply by joulev** [reply]
Agreed. Since we are not yet actually implementing the comments, we don't need to think too far ahead on these exposed variables, but yeah internal gotta be internal.

`_ezk_variableNameInCamelCase`? who forbids us :D
